README.md
# optionee

The macro to create option struct easily.<br>
(no-std support)

[![CI][ci-badge]][ci-url]
[![Crates.io][crates-badge]][crates-url]
[![Licensed][license-badge]][license-url]
[![Twitter][twitter-badge]][twitter-url]

[ci-badge]: https://github.com/just-do-halee/optionee/actions/workflows/ci.yml/badge.svg
[crates-badge]: https://img.shields.io/crates/v/optionee.svg?labelColor=383636
[license-badge]: https://img.shields.io/crates/l/optionee?labelColor=383636
[twitter-badge]: https://img.shields.io/twitter/follow/do_halee?style=flat&logo=twitter&color=4a4646&labelColor=333131&label=just-do-halee

[ci-url]: https://github.com/just-do-halee/optionee/actions
[twitter-url]: https://twitter.com/do_halee
[crates-url]: https://crates.io/crates/optionee
[license-url]: https://github.com/just-do-halee/optionee
| [Docs](https://docs.rs/optionee) | [Latest Note](https://github.com/just-do-halee/optionee/blob/main/CHANGELOG.md) |

```toml
[dependencies]
optionee = "0.2.0"
```

or

```toml
[dependencies]
optionee = { version = "0.2.0", default-features = false } # no-std(alloc)
```

## How to
```rust
optionee! {

    InputOption {

        Id {
            min_length: u8 [>] 2, "id must be more than 3 bytes."
            max_length: u8 [<] 13, "id must be less than 12 bytes."
        }
        Password {
            encrypt: bool [=] true
            min_length: u8 [>] 5, "psasword must be more than 6 bytes."
            max_length: u8 [<] 20, "psasword must be less than 19 bytes."
        }

    }

}

let mut id_t = InputOption.Id();

let user_input = 20;

assert!(id_t.min_length.check(user_input).is_ok());
```
## More Examples
```rust
orderable! {

    pub struct Job {
        id: u32[*], // mark for comparing
        name: String,
        salary: u16[*], // ..
    }

}
optionee! {

    pub TermOption {
            Password {
                max_opportunity: u8 [=] 3, "if you don't really remember your own password, please cider o restart with --reset flag."
                encrypt: bool [=] false
                min_length: u8 [>] 7, "password must be more than 8 lengths bytes."
                max_length: u8 [<] 21, "password must be less than 20 lengths bytes."
            }
    }
    SecondPrivateOpt {
            AnyName {
                name: String [=] "john".to_string()
                job: Job [>] Job::new(0, "sales".to_string(), 29), "id must be more than 1, salary 30"
            }
            OldPerson {
                name: &'static str [=] "mia"
                age: u8 [>] 52
            }

    }

}

let mut t1 = TermOption.Password().encrypt(true).min_length(3);
t1.min_length
    .set_error_message(Some("password must be more than 3 lengths bytes."));

assert!(t1.encrypt.get_value());
assert!(t1.min_length.check(4).is_ok());

let t2 = SecondPrivateOpt
    .AnyName()
    .job(Job::new(0, "writer".to_string(), 29));

assert!(t2.name.check("john".to_string()).is_ok());
assert!(t2.job.check(Job::new(1, "artist".to_string(), 30)).is_ok());
```

CHANGELOG.md
## 0.2.0 (September 5, 2021)

### Release 0.2.0
* anyhow error handling

---

## 0.1.0 (August 24, 2021)

### Release 0.1.0

---

src/macros.rs
// Licensed under either of Apache License, Version 2.0 or MIT license at your option.
// Copyright 2021 Hwakyeom Kim(=just-do-halee)

#[macro_export]
macro_rules! optionee {
    (@transform (>))
    => {
        $crate::private::Ordering::Greater
    };
    (@transform (<))
    => {
        $crate::private::Ordering::Less
    };
    (@transform (=))
    => {
        $crate::private::Ordering::Equal
    };
    (@transform ())
    => {
        None
    };
    (@transform ($e:expr))
    => {
        Some($e)
    };
    (
        $(
            $vis:vis $id:ident {
                $(
                    $name:ident {
                        $(
                            $var:ident: $t:ty [$cmp:tt] $val:expr $(, $err:expr)?
                        )+
                    }
                )+
            }
        )+
    ) => {
        $(
            pub struct $id;
            impl $id {
                $(
                    #[allow(non_snake_case, dead_code)]
                    pub fn $name(self) -> optionees::$id::$name {
                        optionees::$id::$name::new()
                    }
                )+
            }
        )+
        mod optionees {
            #![allow(unused_imports)]
            #![allow(dead_code)]
            use super::*;
            use $crate::private::*;
            pub struct Wrapping<T: Ord>(T, Ordering, Option<&'static str>);
            impl<T: Ord> Wrapping<T> {
                pub fn check(&self, value: T) -> Result<(), Error> {
                    let ordering = self.1;
                    let msg = self.2;
                    if value.cmp(&self.0) != ordering {
                        if let Some(v) = msg {
                            Err(Error::msg(v))
                        } else {
                            Err(Error::msg("not matched."))
                        }
                    } else {
                        Ok(())
                    }
                }
                pub fn get_value(&self) -> &T {
                    &self.0
                }
                pub fn get_error_message(&self) -> Option<&'static str> {
                    self.2
                }
                pub fn set_error_message(&mut self, msg: Option<&'static str>) {
                    self.2 = msg;
                }
            }
            $(
                #[allow(non_snake_case)]
                pub mod $id {
                    use super::*;
                    $(
                        pub struct $name {
                            $(
                                pub $var: Wrapping<$t>
                            ),+
                        }
                        impl $name {
                            pub fn new() -> Self {
                                Self {
                                    $(
                                        $var: Wrapping($val, optionee!(@transform ($cmp)), optionee!(@transform ($($err)?)))
                                    ),*
                                }
                            }
                            $(
                                pub fn $var(mut self, value: $t) -> Self {
                                    self.$var.0 = value;
                                    self
                                }
                            )*
                        }
                    )+
                }
            )+
        }
    };
}

#[macro_export]
macro_rules! orderable {
    (@transform [])
    => {
        false
    };
    (@transform [*])
    => {
        true
    };
    (
        $(#[$meta:meta])*
        $vis:vis struct $name:ident {
            $(
                $(#[$field:meta])*
                $var:ident: $t:ty$([$tt:tt])?$(,)?
            )+
        }
    ) => {
        $(#[$meta])*
        #[derive(Eq)]
        $vis struct $name {
            $(
                $(#[$field])*
                $var: $t
            ),+
        }
        impl $name {
            pub fn new($($var: $t),+) -> Self {
                Self {
                    $(
                        $var
                    ),+
                }
            }
        }
        impl $crate::private::Ord for $name {
            fn cmp(&self, other: &Self) -> $crate::private::Ordering {
                let mut all = alloc::vec![];
                $(
                    if orderable!(@transform [$($tt)?]) {
                        all.push(self.$var.cmp(&other.$var))
                    }
                )*
                let first = all[0];
                if !all.iter().all(|&item| item == first) {
                    panic!("not matched all items' ordering.");
                }
                first
            }
        }
        impl $crate::private::PartialOrd for $name {
            fn partial_cmp(&self, other: &Self) -> $crate::private::Option<$crate::private::Ordering> {
                Some(self.cmp(other))
            }
        }
        impl $crate::private::PartialEq for $name {
            fn eq(&self, other: &Self) -> bool {
                $(self.$var == other.$var)&&*
            }
        }
    };
}


src/lib.rs
// Licensed under either of Apache License, Version 2.0 or MIT license at your option.
// Copyright 2021 Hwakyeom Kim(=just-do-halee)

//! # `Optionee`
//!
//! The macro to create option struct easily.<br>
//! (no-std support)
//! ## How to
//! ```rust
//! # #[macro_use] extern crate optionee; extern crate alloc;
//! # use optionee::*;
//! optionee! {
//!     InputOption {
//!         Id {
//!             min_length: u8 [>] 2, "id must be more than 3 bytes."
//!             max_length: u8 [<] 13, "id must be less than 12 bytes."
//!         }
//!         Password {
//!             encrypt: bool [=] true
//!             min_length: u8 [>] 5, "psasword must be more than 6 bytes."
//!             max_length: u8 [<] 20, "psasword must be less than 19 bytes."
//!         }
//!     }
//! }
//! # fn main() {
//! let mut id_t = InputOption.Id();
//! let user_input = 20;
//! assert!(id_t.min_length.check(user_input).is_ok());
//! # }
//! ```
//!
//! ## More Examples
//! ```rust
//! # #[macro_use] extern crate optionee; extern crate alloc;
//! # use optionee::*;
//! orderable! {
//!     pub struct Job {
//!         id: u32[*],
//!         name: String,
//!         salary: u16[*],
//!     }
//! }
//!
//! optionee! {
//!     pub TermOption {
//!             Password {
//!                 max_opportunity: u8 [=] 3, "if you don't really remember your own password, please consider to restart with --reset flag."
//!                 encrypt: bool [=] false
//!                 min_length: u8 [>] 7, "password must be more than 8 lengths bytes."
//!                 max_length: u8 [<] 21, "password must be less than 20 lengths bytes."
//!             }
//!         }
//!     SecondPrivateOpt {
//!             AnyName {
//!                 name: String [=] "john".to_string()
//!                 job: Job [>] Job::new(0, "sales".to_string(), 29), "id must be more than 1, salary 30"
//!             }
//!             OldPerson {
//!                 name: &'static str [=] "mia"
//!                 age: u8 [>] 52
//!             }
//!         }
//! }
//! # fn main() {
//! let mut t1 = TermOption.Password().encrypt(true).min_length(3);
//! t1.min_length
//!     .set_error_message(Some("password must be more than 3 lengths bytes."));
//!
//! assert!(t1.encrypt.get_value());
//! assert!(t1.min_length.check(4).is_ok());
//!
//! let t2 = SecondPrivateOpt
//!     .AnyName()
//!     .job(Job::new(0, "writer".to_string(), 29));
//!
//! assert!(t2.name.check("john".to_string()).is_ok());
//! assert!(t2.job.check(Job::new(1, "artist".to_string(), 30)).is_ok());
//! # }
//! ```

#![cfg_attr(all(not(feature = "std"), not(test)), no_std)]

pub mod private {
    pub use anyhow::Error;
    #[cfg(not(feature = "std"))]
    pub use core::{
        cmp::{Ord, Ordering, PartialEq, PartialOrd},
        option::Option,
    };
    #[cfg(feature = "std")]
    pub use std::{
        cmp::{Ord, Ordering, PartialEq, PartialOrd},
        option::Option,
    };
}

extern crate alloc;

#[macro_use]
mod macros;

#[cfg(test)]
mod tests {
    use super::*;
    use std::string::ToString;

    orderable! {
        pub struct Job {
            id: u32[*],
            name: String,
            salary: u16[*],
        }
    }

    optionee! {
        pub TermOption {
                Password {
                    max_opportunity: u8 [=] 3, "if you don't really remember your own password, please consider to restart with --reset flag."
                    encrypt: bool [=] false
                    min_length: u8 [>] 7, "password must be more than 8 lengths bytes."
                    max_length: u8 [<] 21, "password must be less than 20 lengths bytes."
                }
            }
        SecondPrivateOpt {
                AnyName {
                    name: String [=] "john".to_string()
                    job: Job [>] Job::new(0, "sales".to_string(), 29), "id must be more than 1, salary 30"
                }
                OldPerson {
                    name: &'static str [=] "mia"
                    age: u8 [>] 52
                }
            }
    }
    #[test]
    fn it_works() {
        let mut t1 = TermOption.Password().encrypt(true).min_length(3);
        t1.min_length
            .set_error_message(Some("password must be more than 3 lengths bytes."));

        assert!(t1.encrypt.get_value());
        assert!(t1.min_length.check(4).is_ok());

        let t2 = SecondPrivateOpt
            .AnyName()
            .job(Job::new(0, "writer".to_string(), 29));

        assert!(t2.name.check("john".to_string()).is_ok());
        assert!(t2.job.check(Job::new(1, "artist".to_string(), 30)).is_ok());
    }
}
